homeassistant/components/stream/hls.py

```diff
@@ -187,28 +321,91 @@ class HlsSegmentView(StreamView):
-        """Return provider name."""
-        return HLS_PROVIDER
+        # Ensure that we have a segment. If the request is from a hint for part 0
+        # of a segment, there is a small chance it may have arrived before the
+        # segment has been put. If this happens, wait for one part and retry.
+        if not (
+            (segment := track.get_segment(int(sequence)))
+            or (
+                await track.part_recv(timeout=StreamSettings.hls_part_timeout)
+                and (segment := track.get_segment(int(sequence)))
+            )
+        ):
+            return web.Response(
+                body=None,
+                status=404,
+                headers={"Cache-Control": f"max-age={track.target_duration:.0f}"},
+            )
+        # If the segment is ready or has been hinted, the http_range start should be at most
+        # equal to the end of the currently available data.
+        # If the segment is complete, the http_range start should be less than the end of the
+        # currently available data.
+        # If these conditions aren't met then we return a 416.
+        # http_range_start can be None, so use a copy that uses 0 instead of None
+        if (http_start := request.http_range.start or 0) > segment.data_size or (
+            segment.complete and http_start >= segment.data_size
+        ):
+            return web.HTTPRequestRangeNotSatisfiable(
+                headers={
+                    "Cache-Control": f"max-age={track.target_duration:.0f}",
+                    "Content-Range": f"bytes */{segment.data_size}",
+                }
+            )
+        headers = {
+            "Content-Type": "video/iso.segment",
+            "Cache-Control": f"max-age={6*track.target_duration:.0f}",
+        }
+        # For most cases we have a 206 partial content response.
+        status = 206
+        # For the 206 responses we need to set a Content-Range header
+        # See https://datatracker.ietf.org/doc/html/rfc8673#section-2
+        if request.http_range.stop is None:
+            if request.http_range.start is None:
+                status = 200
+                if segment.complete:
+                    # This is a request for a full segment which is already complete
+                    # We should return a standard 200 response.
+                    return web.Response(
+                        body=segment.get_data(), headers=headers, status=status
+                    )
+                # Otherwise we still return a 200 response, but it is aggregating
+                http_stop = float("inf")
+            else:
+                # See https://datatracker.ietf.org/doc/html/rfc7233#section-2.1
+                headers[
+                    "Content-Range"
+                ] = f"bytes {http_start}-{(http_stop:=segment.data_size)-1}/*"
+        else:  # The remaining cases are all 206 responses
+            if segment.complete:
+                # If the segment is complete we have total size
```

Comment:
For my understanding (I didn't read any rfcs): The segment is complete but this is still returned as a 206 partial response?

Reply:
Yes, I think if we get a `Range` request we reply with a 206. Note that we are just checking if `segment.complete` in order to fill in the total byte size, but we are not sending the whole segment.